lib/facter/powershell_version.rb

```diff
@@ -0,0 +1,76 @@
+Facter.add(:powershell_version) do
+  setcode do
+    if Puppet::Util::Platform.windows?
+      version = Facter.value(:powershell_three_version) || Facter.value(:powershell_one_version)
+    end
+  end
+end
+
+Facter.add(:powershell_one_version) do
+  setcode do
+    if Puppet::Util::Platform.windows?
+      require 'win32/registry'
+      access_type = Win32::Registry::KEY_READ | 0x100
+      hklm        = Win32::Registry::HKEY_LOCAL_MACHINE
+      reg_path    = 'SOFTWARE\Microsoft\PowerShell\1\PowerShellEngine'
+      reg_key     = 'PowerShellVersion'
+      version     = nil
+      begin
+        hklm.open(reg_path, access_type) do |reg|
+          version = reg[reg_key]
+        end
+      rescue
+        version = nil
+      end
+    end
+    version
+  end
+end
+
+Facter.add(:powershell_three_version) do
+  setcode do
+    if Puppet::Util::Platform.windows?
+      require 'win32/registry'
+      access_type = Win32::Registry::KEY_READ | 0x100
+      hklm        = Win32::Registry::HKEY_LOCAL_MACHINE
+      reg_path    = 'SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine'
+      reg_key     = 'PowerShellVersion'
+      version     = nil
+      begin
+        hklm.open(reg_path, access_type) do |reg|
+          version = reg[reg_key]
+        end
+      rescue
+        version = nil
+      end
+    end
+    version
+  end
+end
+
+Facter.add(:powershell_major_version) do
```

Comment:
If we return a `Gem::Version` as `:powershell_version` then I don't think these bits are necessary.